cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -0,0 +1,867 @@
+        if(is_lanechange_accepted_){
+            cav_msgs::TrajectoryPlan trajectory_plan;
+            trajectory_plan.header.frame_id = "map";
+            trajectory_plan.header.stamp = ros::Time::now();
+            trajectory_plan.trajectory_id = boost::uuids::to_string(boost::uuids::random_generator()());
+
+            trajectory_plan.trajectory_points = planned_trajectory_points;
+            trajectory_plan.initial_longitudinal_velocity = std::max(req.vehicle_state.longitudinal_vel, minimum_speed_);
+            resp.trajectory_plan = trajectory_plan;
+
+            for (int i=0; i<req.maneuver_plan.maneuvers.size(); i++){
+                if (req.maneuver_plan.maneuvers[i].type == cav_msgs::Maneuver::LANE_CHANGE){
+                    resp.related_maneuvers.push_back(i);
+                    break;
+                }
+            }
+            resp.maneuver_status.push_back(cav_srvs::PlanTrajectory::Response::MANEUVER_IN_PROGRESS);
+
+        }
+        else{
+            ros::Time planning_end_time = ros::Time::now();
+            ros::Duration passed_time = planning_end_time - request_sent_time;
+            if(passed_time.toSec() >= lanechange_time_out_){
+                cav_msgs::LaneChangeStatus lc_status_msg;
+                lc_status_msg.status = cav_msgs::LaneChangeStatus::TIMED_OUT;
+                lc_status_msg.description = "Request timed out for lane merge";
+                lanechange_status_pub_.publish(lc_status_msg);
+            }
+ 
+        }
+
+        return true;
+    }
+    
+    cav_msgs::MobilityRequest CooperativeLaneChangePlugin::create_mobility_request(std::vector<cav_msgs::TrajectoryPlanPoint>& trajectory_plan, cav_msgs::Maneuver maneuver){
+
+        cav_msgs::MobilityRequest request_msg;
+        cav_msgs::MobilityHeader header;
+        header.sender_id = sender_id_;
+        header.recipient_id = DEFAULT_STRING_;      //To do- find way to query recipient id
+        header.sender_bsm_id = bsmIDtoString(bsm_core_);
+        header.plan_id = boost::uuids::to_string(boost::uuids::random_generator()());
+        header.timestamp = trajectory_plan.front().target_time.toSec();
+        request_msg.header = header;
+
+        request_msg.strategy = "carma/cooperative-lane-change";
+        request_msg.plan_type.type = cav_msgs::PlanType::CHANGE_LANE_LEFT;
+        //Urgency- Currently unassigned
+
+        //Location
+        cav_msgs::LocationECEF location;
+        location.ecef_x = pose_msg_.pose.position.x;
+        location.ecef_y = pose_msg_.pose.position.y;
+        location.ecef_z = pose_msg_.pose.position.z;
+        //Using trajectory first point time as location timestamp;
+        location.timestamp = trajectory_plan.front().target_time.toSec();
+        request_msg.location = location;
+        //Strategy params
+        //Encode JSON with Boost Property Tree
+        using boost::property_tree::ptree;
+        ptree pt;
+        pt.put("speed",maneuver.lane_change_maneuver.start_speed);
```

Comment:
This needs to be considered carefully. I was talking with @saina-ramyar about exactly how we're determining the parameters for the jerk-minimizing-trajectory used to cause the yielding vehicle to slow down. You should check with her to make sure we're keeping this two aligned in terms of how we're coordinating this gap creation.